providers/ldap/ad_provider.go

```diff
@@ -0,0 +1,245 @@
+package ldap
+
+import (
+	"net/http"
+	"strconv"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/rancher/go-rancher/v2"
+	"github.com/rancher/rancher-auth-service/model"
+)
+
+const (
+	Name                            = "ldap"
+	UserScope                       = Name + "_user"
+	GroupScope                      = Name + "_group"
+	Config                          = Name + "config"
+	LdapJwt                         = Name + "Jwt"
+	MemberOfAttribute               = "memberOf"
+	ObjectClassAttribute            = "objectClass"
+	settingBase                     = "api.auth.ldap."
+	AccessModeSetting               = settingBase + "access.mode"
+	DomainSetting                   = settingBase + "domain"
+	GroupSearchDomainSetting        = settingBase + "group.search.domain"
+	LoginDomainSetting              = settingBase + "login.domain"
+	PortSetting                     = settingBase + "port"
+	UserSearchFieldSetting          = settingBase + "user.search.field"
+	ServiceAccountUsernameSetting   = settingBase + "service.account.user"
+	GroupSearchFieldSetting         = settingBase + "group.search.field"
+	UserObjectClassSetting          = settingBase + "user.object.class"
+	UserNameFieldSetting            = settingBase + "user.name.field"
+	GroupObjectClassSetting         = settingBase + "group.object.class"
+	UserLoginFieldSetting           = settingBase + "user.login.field"
+	UserDisabledBitMaskSetting      = settingBase + "user.enabled.mask.bit"
+	ServerSetting                   = settingBase + "server"
+	ServiceAccountPasswordSetting   = settingBase + "service.account.password"
+	UserEnabledAttributeSetting     = settingBase + "user.enabled.attribute"
+	GroupNameFieldSetting           = settingBase + "group.name.field"
+	TlsSetting                      = settingBase + "tls"
+	TimeoutSetting                  = settingBase + "connection.timeout"
+	AllowedIdentitiesSetting        = settingBase + "allowed.identities"
+	GroupDnFieldSetting             = settingBase + "group.dn.field"
+	GroupMemberUserAttributeSetting = settingBase + "group.member.user.attribute"
+)
+
+var scopes = []string{UserScope, GroupScope}
+
+type ADConstantsConfig struct {
+	userScope            string
+	groupScope           string
+	scopes               []string
+	memberOfAttribute    string
+	objectClassAttribute string
+}
+
+func InitializeProvider() *LdapProvider {
+	client := &http.Client{}
+	ldapClient := &LClient{}
+	ldapClient.httpClient = client
+
+	ldapProvider := &LdapProvider{}
+	ldapProvider.ldapClient = ldapClient
+
+	adConstantsConfig := &ADConstantsConfig{
+		userScope:            UserScope,
+		groupScope:           GroupScope,
+		scopes:               scopes,
+		memberOfAttribute:    MemberOfAttribute,
+		objectClassAttribute: ObjectClassAttribute,
+	}
+	ldapProvider.ConstantsConfig = adConstantsConfig
+
+	return ldapProvider
+}
+
+type LdapProvider struct {
```

Comment:
Will OpenLDAP have its own OpenLDAPProvider struct? If so can we rename this to ADProvider for readability?